roles/lib_openshift/src/class/oc_configmap.py

```diff
@@ -0,0 +1,183 @@
+                 name,
+                 from_file,
+                 from_literal,
+                 state,
+                 namespace,
+                 kubeconfig='/etc/origin/master/admin.kubeconfig',
+                 verbose=False):
+        ''' Constructor for OpenshiftOC '''
+        super(OCConfigMap, self).__init__(namespace, kubeconfig=kubeconfig, verbose=verbose)
+        self.name = name
+        self.state = state
+        self._configmap = None
+        self._inc_configmap = None
+        self.from_file = from_file if from_file is not None else {}
+        self.from_literal = from_literal if from_literal is not None else {}
+
+    @property
+    def configmap(self):
+        if self._configmap is None:
+            self._configmap = self.get()
+
+        return self._configmap
+
+    @configmap.setter
+    def configmap(self, inc_map):
+        self._configmap = inc_map
+
+    @property
+    def inc_configmap(self):
+        if self._inc_configmap is None:
+            results = self.create(dryrun=True, output=True)
+            self._inc_configmap = results['results']
+
+        return self._inc_configmap
+
+    @inc_configmap.setter
+    def inc_configmap(self, inc_map):
+        self._inc_configmap = inc_map
+
+    def from_file_to_params(self):
+        '''return from_files in a string ready for cli'''
+        return ["--from-file={}={}".format(key, value) for key, value in self.from_file.items()]
+
+    def from_literal_to_params(self):
+        '''return from_literal in a string ready for cli'''
+        return ["--from-literal={}={}".format(key, value) for key, value in self.from_literal.items()]
+
+    def get(self):
+        '''return a configmap by name '''
+        results = self._get('configmap', self.name)
+        if results['returncode'] == 0 and results['results'][0]:
+            self.configmap = results['results'][0]
+
+        if results['returncode'] != 0 and '"{}" not found'.format(self.name) in results['stderr']:
+            results['returncode'] = 0
+
+        return results
+
+    def delete(self):
+        '''delete a configmap by name'''
+        return self._delete('configmap', self.name)
+
+    def create(self, dryrun=False, output=False):
+        '''Create a configmap
+
+           :dryrun: Product what you would have done. default: False
+           :output: Whether to parse output. default: False
+        '''
+
+        cmd = ['create', 'configmap', self.name]
+        if self.from_literal is not None:
+            cmd.extend(self.from_literal_to_params())
+
+        if self.from_file is not None:
+            cmd.extend(self.from_file_to_params())
+
+        if dryrun:
+            cmd.extend(['--dry-run', '-ojson'])
+
+        results = self.openshift_cmd(cmd, output=output)
+
+        return results
+
+    def update(self):
+        '''run update configmap '''
+        return self._replace_content('configmap', self.name, self.inc_configmap)
+
+    def needs_update(self):
```

Comment:
@stevekuznetsov, Thanks for the conversation.  The `oc patch` has the same behavior of `oc apply`.
```bash
[root@ip-172-31-60-178 ~]# oc patch dc/mysql -p '{"spec":{"replicas": 2}}'
"mysql" patched
[root@ip-172-31-60-178 ~]# echo $?
0
[root@ip-172-31-60-178 ~]# oc patch dc/mysql -p '{"spec":{"replicas": 2}}'
"mysql" patched
[root@ip-172-31-60-178 ~]# echo $?
0
```
If we can push this behavior upstream so that we can reliably determine changes actually occurred I'm fully on board with updating to match those tools.

The other issue I have with patch is that it takes slices of an item.  Most of the time we want to manage state and when doing so we want multiple edits to occur to an object.

As for the example, `oc patch node k8s-node-1 -p '{"spec":{"unschedulable":true}}'` this requires passing a `-p` and a specific or `strategic` slice into the file.  I did not research more into this per the issue of it always resulting in a returncode of 0 and saying, "patched".

Does patch also allow multiple cuts as a single transaction?  Does it take an entire object?  Until then, I think the `oc replace` for an update, which does a `put` under the covers is exactly what we want.  Its a single transaction and it we get the behavior that we want.

I'm definitely open for discussion when it comes to architecting the modules and improving them.  I'm not sold on `PATCH` over `oc replace`.